Mesh: Add differentiable shrinkwrap deformation (#1886)
NVIDIA/physicsnemo: docs/api/nn/functionals/geometry.rst

path. Because its backward uses atomic accumulation at shared vertices, Warp
gradient results can have small run-to-run floating-point differences.

> Nearest-Surface Shrinkwrap

Reviewer: I doesn't look like shrinkwrap (the operation) is defined here; can we add a definition for users?

--------------------------

.. autofunction:: physicsnemo.nn.functional.shrinkwrap_points

.. code:: python

    import torch
    from physicsnemo.nn.functional import shrinkwrap_points

    target_points = torch.tensor(
        [
            [0.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 1.0, 0.0],
            [0.0, 1.0, 0.0],
        ],
        requires_grad=True,
    )
    target_faces = torch.tensor([[0, 1, 2], [0, 2, 3]])
    points = torch.tensor(
        [[0.25, 0.25, 0.30], [0.75, 0.75, 0.45]],
        requires_grad=True,
    )
    weights = torch.tensor([0.5, 1.0], requires_grad=True)
    offset = torch.tensor(0.02, requires_grad=True)

    wrapped = shrinkwrap_points(
        points,
        target_points,
        target_faces,
        point_weights=weights,
        offset=offset,
    )
    wrapped.square().mean().backward()

> Source points can have shape ``(N, 3)`` or ``(B, N, 3)``. One triangle target

Reviewer 2: ```suggestion
Source points use one of these shapes:

- ``(N, 3)``
- ``(B, N, 3)``

One triangle target is shared across the source batches. Keep the following
projection rules in mind:

- A positive ``offset`` follows target face winding.
- A finite ``max_distance`` leaves points at or beyond the cutoff unchanged.
 
```

Reviewer 2: then delete lines 325-327

is shared across the source batches. A positive ``offset`` follows target face
winding. A finite ``max_distance`` leaves points at or beyond the cutoff
unchanged.

Nearest-face selection, closest-feature changes, and distance gating are

> discrete. Between those transitions, gradients propagate through source

Reviewer 2: ```suggestion
discrete. Between those transitions, gradients propagate through the following
values:

- Source points
- Selected target vertices
- Floating-point weights
- A tensor-valued ``offset``

Target connectivity is not differentiable.
```

> points, selected target vertices, floating point weights, and a tensor-valued

Reviewer 2: ```suggestion
```

> ``offset``. Target connectivity is not differentiable.

Reviewer 2: ```suggestion
```

Torch provides the reference nearest-face search. Warp accelerates that search
on CPU and CUDA. Both backends evaluate the selected projection with PyTorch

> in the input dtype. Exact ties at target edges or vertices can select different

Reviewer 2: on the fence about whether dtype should be formatted as ``dtype`` or not.  Prevailing advice would be to stay consistent within this file, so I will leave that as is.

adjacent faces. Their closest point is the same, but a nonzero face-normal
offset can differ.

> Float64 targets use the Torch search because Warp searches in float32. Safe

Reviewer 2: ```suggestion
Search backend rules differ by dtype and geometry:

- ``float64`` targets use the Torch search because Warp searches in ``float32``.
- Warp searches safe ``float32`` coordinates unchanged.
- Warp falls back to Torch for unsafe coordinate magnitudes or face geometry.
```

float32 coordinates are searched unchanged. Warp falls back to Torch for
unsafe coordinate magnitudes or face geometry.

Shrinkwrap performs data-dependent validation and nearest-face search setup.

> CUDA executions with either backend are not supported inside CUDA Graph

Reviewer 2: ```suggestion
Do not run CUDA executions with either backend inside CUDA Graph capture.

```

> capture.

Reviewer 2: ```suggestion
```
